src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs

```diff
@@ -215,20 +248,69 @@ public T GetValue(CancellationToken cancellationToken)
             // but we don't want multiple threads attempting to compute the same thing.
             if (request != null)
             {
+                // Always want to make sure that if we ourselves are cancelling this request, that the request is cleaned up.
                 request.RegisterForCancellation(OnAsynchronousRequestCancelled, cancellationToken);
 
-                // Since we already registered for cancellation, it's possible that the registration has
-                // cancelled this new computation if we were the only requester.
-                if (newAsynchronousComputation != null)
+                if (forceSynchronousComputeOnCallingThread)
                 {
-                    StartAsynchronousComputation(newAsynchronousComputation.Value, requestToCompleteSynchronously: request, callerCancellationToken: cancellationToken);
+                    // Should never get here since it requires _synchronousComputeFunction be null, and that's the 
+                    // first thing this method checks for.
+                    Contract.ThrowIfTrue(newAsynchronousComputation != null);
+                    Contract.ThrowIfNull(_synchronousComputeFunction);
+
+                    // Caller wants to try to compute on their own thread.  But we also have a real in-flight
+                    // computation going on in another thread that may win.  We want to do the work on our thread, but
+                    // also cancel that work if the other thread wins.  To that effect, we make a linked token to
+                    // represent our work.
+                    using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+                    // If the async work finishes, cancel the current work on this thread.
+                    request.Task.ContinueWith(_ => linkedTokenSource.Cancel(), CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
```

Comment:
Open question if we should have TaskCOntinuationOptions.RunSynchronously.

Reply:
My assumption was yes: otherwise we could have the result get computed on the thread pool, but the synchronous result _still_ won't get used since we're just waiting for this .Cancel() to run.